5.7.37 psmerge part 1

nitendra-bhosle opened the pull request.

patrickbirch committed:
PS-3313 When using proxy protocol, mysql must bind to an address instead of 0.0.0.0

modified file:
flexibility/proxy_protocol_support

Rewrote first paragraph
added note to the variable

patrickbirch committed:
PS-3740 LP #1716507: innodb_flush_method needs updating in documentation

modified files:
scalability/innodb_io
changed the default value
changed the allowed values to match 5.6

pedrolgomes committed:
PS-7792: On replica Master_User should be blank if not storing in slave_master_info
BUG#27357189: MASTER_USER BECOMES 'TEST' IF IT IS EMPTY WHEN READ FROM REPO

This commit backports the bugfix for Oracle Bug#27357189 to PS-5.7.

When the user for the master connection is empty and there is a read
from the repository information the user becomes 'test'. This happens
as the default value when there is no information is 'test'.

The default value is now empty as well.

ReviewBoard: 24194
(cherry picked from commit 304e920403cecaab9a45834d041bc2ea5a0d2a57)

Commit:
PS-8030: Multithreaded replica crashes inside Relay_log_info::cannot_safely_rollback()

https://jira.percona.com/browse/PS-8030

This issue backports PS-7147 (MTA Replica crashes inside
Transaction_ctx::THD_TRANS::cannot_safely_rollback())

(https://jira.percona.com/browse/PS-7147)

Problem & Analysis
------------------
As part of the bugfix for PS-5824 (commit 205b48f017a9842c736d3), the
co-ordinator thread was made to check for the safeness of the ongoing
transactions and if found that all transactions executed by the worker
threads are safe to rollback, then it immediately would stop all the worker
threads making the STOP SLAVE to execute faster.

However, while checking for the safeness, it doesn't take into account that
the transaction objects can be empty if the worker thread had performed the
clean up of the transaction objects.

This caused the the function call to
`worker->info_thd->get_transaction()->cannot_safely_rollback(Transaction_ctx::SESSION);`
result in dereferencing a null pointer and caused the mysqld to crash.

Fix
---
Added a check for null pointer.

(cherry picked from 8.0 commit b68dffffa46ab16f94da7f380cba492d48041f11)

karthik-kamath committed:
BUG#31774422: SERVER HANGS IF BINLOG ROTATION CAUSED BY
              VARIABLE UPDATE + CONCURRENT READ

PROBLEM:
========
Transaction commit during update of the system variable
which resulted in binlog rotation, while concurrently
reading variables from another connection might have led to
deadlock.

ANALYSIS:
=========
When connection commits transaction during update of the
system variable it is done after taking THD::LOCK_thd_sysvar
lock. During this commit, binlog rotation can happen, which
will cause THD::LOCK_thd_data lock being acquired when
opening gtid_executed table to write the executed GTIDs.
Code which reads value of system variables for the first
connection from another connection will first take
THD::LOCK_thd_data of the first thread (to prevent it from
going away) and later take its THD::LOCK_thd_sysvar (to
prevent it from updating system variable concurrently).

This opposite lock order in two connections causes a
deadlock.

FIX:
====
Introduced a pre_update hook as a preparatory step for
updating a system variable to be executed before acquiring
THD::LOCK_sys_var. Transaction commit happens on this step
without any lock held and thus we can safely acquire
THD::LOCK_thd_data without creating problems with lock
order.

RB#25744

Change-Id: I415f5bae06ff105f1af83b53dc672f3addb92235

patrickbirch committed:
Merge branch '5.6' into ps-5.6-7514

patrickbirch committed:
Merge pull request #4194 from patrickbirch/ps-5.6-7514

PS-7514 Release notes for Percona Server for MySQL (5.6)

patrickbirch committed:
PS-7554 Additional text missing

modified:   source/release-notes/Percona-Server-5.6.51-91.0.rst

patrickbirch committed:
Merge pull request #4196 from patrickbirch/ps-5.6-7554

PS-7554 Additional text missing (5.6)

patrickbirch committed:
PS-7555 Update 5.6 release notes (5.6)

modified:   source/release-notes/Percona-Server-5.6.51-91.0.rst

patrickbirch committed:
Merge pull request #4197 from patrickbirch/ps-5.6-7555

PS-7555 Update 5.6 release notes (5.6)

patrickbirch committed:
Merge pull request #3588 from patrickbirch/ps-5.6-4641

PS-4641 innodb_flush_method is not clear enough

patrickbirch committed:
Merge pull request #3634 from patrickbirch/ps-5.6-3740

PS-3740 LP #1716507: innodb_flush_method needs updating in documentation

patrickbirch committed:
Merge pull request #3630 from patrickbirch/ps-5.6-3313

PS-3313 When using proxy protocol, mysql must bind to an address inst…

patrickbirch committed:
Merge pull request #3642 from patrickbirch/ps-5.6-4670

PS-4670 fix COMMAND_CLASS text

patrickbirch committed:
PS-7600 Non working wget command on the documentation (5.6)

 Changes to be committed:
	modified:   source/upgrading_guide_55_56.rst

patrickbirch committed:
Merge pull request #4233 from patrickbirch/ps-5.6-7600

PS-7600 Non working wget command on the documentation (5.6)

Commit:
PS-7746 - Assign NULL to ha_innobase::m_share when free_share

Specific PS code might expect `share` to be valid when it is not NULL.

*Solution*: After any call to `free_share`, NULL is assigned to `share`.

patrickbirch committed:
PS-7779 Document system-wide crypto-policies for RHEL/CentOS (5.7)

 Changes to be committed:
	modified:   source/installation/yum_repo.rst

patrickbirch committed:
PXC-3699 wsrep_cluster_name documentation wrong about dynamic option (5.6)

Modified wsrep-system-index

patrickbirch committed:
Merge pull request #4394 from patrickbirch/ps-5.6-7795

PS-7795 Update docs for log_slow_verbosity = standard (5.6)

avbelov23 committed:
Bug #32316323 MYSQL.EXE SILENTLY CRASHED ON WINDOWS 10 ZIPPED ARCHIVE

Fixed Windows specific character count in buffer miscalculation in
my_win_console_write.

ReviewBoard: 25735

avbelov23 committed:
Bug #31899685 INNODB HANDLER SHARE'S INDEX_MAPPING ACCESS INVALID

MEMORY OF DICT INDEX

Problem:
Innodb table share instance (m_share) and dict table instance
open/close sequence is:
  get_share()
    open_dict_table()
    close_dict_table()
  free_share()
In rare case, access to m_share object cause segment fault or hit
an assert.

For Example: Transaction T1 & T2 access table A in following
sequence -
1. T1 get an m_share object of table A.
2. Just after step 1, If T2 close table A and evict its
dict_table_t instance from dict cache.
3. Now if T1 re-open table A then m_share instance is
old and point to invalid dict indexes.
In this case, if we try to access m_share object will cause
segment fault or hit an assert.

Fix:
Modified Innodb table share instance (m_share) and dict table
instance open/close sequence to:
  open_dict_table()
    get_share()
    free_share()
  close_dict_table()
This will protect dict index instances in m_share to be evicted
from dict cache.

RB: 25786
Reviewed by : Debarun Banerjee <debarun.banerjee@oracle.com>

avbelov23 committed:
Bug #31674599: THE UDF_INIT() FUNCTION CAUSE SERVER CRASH.

Description:
Fix an issue in udf_init() so that it can throw error for
wrong name.

Change-Id: I81d7954837470b3dfd991e7da4f7ba79b9661977

avbelov23 committed:
BUG#30366310: USING A FUNCTION TO ASSIGN DEFAULT VALUES TO

              2 OR MORE VARIABLES CRASHES SERVER

Description:
============
Consider SP declaring multiple local variables in the same
DECLARE statement with expression as value in DEFAULT
clause. Attempting to call such SP more than once could
result in server exit.

Analysis:
=========
If multiple variables are declared in the same DECLARE
statement in SP, then internally they are converted to
multiple SET instructions. All the SET instructions will
have shared default value expression. But each instruction
in SP is supposed to be self-contained, with its own item
list.

While executing first SET instruction, items in instructions
are added to thd->free_list and it is cleaned up at the end
of instruction execution. But in the subsequent execution,
default expression items are not added to the
thd->free_list, so default expression is not cleaned up
properly. Executing SP again resulted in invalid access.

Fix
===
Even SET instructions of SP are made self-contained to fix
this issue. Cloning default value expression for every
subsequent declaration is one of the possible solution but
it is not feasible. So to make SET instruction
self-contained, second assignment is replaced with
assignment from first. So we are internally transforming
'DECLARE a, b, c type DEFAULT expr;'
 to
'DECLARE a type DEFAULT expr, b type DEFAULT a,
         c type DEFAULT a;'.
Such transformation is done for only non-const default
expressions.

Original patch by Gleb Shchepa.

Change-Id: I522aabc201930a5a7844a07fa1aee63b66c5aacd

avbelov23 committed:
Bug #32391415: MYSQL_CHANGE_USER FAILS WITH UNKNOWN COLLATION

RB#26447

mysql_change_user() in 5.7 was not properly parsing the COM_CHANGE_USER
packet and this results in a quiet failure to process the optional query
attributes that may have been supplied pre- mysql_change_user().

Fixed to properly process the packet and properly extract the query
attributes.
Test case added.

Note that this issue is 5.7 specific since in 8.0 the query attributes
are parsed properly.

avbelov23 committed:
Bug #31576731 INNODB_FT_TOTAL_CACHE_SIZE NOT CAPPED WHEN SET TO

32000000

Problem:
Server throws OOM error when we execute twitter load with SELECTs
for UPDATE + UPDATES, and SELECT queries on tables with full-text
index.
FTS cache->total_memory store count of total memory allocated to FTS
cache (for all fulltext indexes on a table).
For each word in fts cache, we store doc-id & word position in a
node->ilist.
we increment cache->total_memory with size of doc-id & word position
whereas we allocate ilist in chuck of 16, 32 ,64 bytes or 1.2 times
of last size.
When we wil insert huge amount of data into the FTS aux index tables
then collectively these small chucks for each token become huge
unaccounted memory allocated for FTS cache.

Fix:
Incremented cache->total_memory by size of chunk allocated to
node->ilist.

RB: 25286
Reviewed by : Rahul Agarkar <rahul.agarkar@oracle.com>

avbelov23 committed:
Bug#32620378, Bug#32620398: ZDI-CAN-13265, ZDI-CAN-13266: MYSQL MEMCA…

…CHED PLUGIN INTEGER UNDERFLOW

Fixed integer underflow possibilities of memcached code in innodb_memcached plugin

RB: 26411
Reviewed-by: Sunny Bains <Sunny.Bains@oracle.com>

kamil-holubicki committed:
PS-7620: 5.7.34 and 5.7.35 backports

Version bumped

kamil-holubicki committed:
Merge pull request #4422 from tempesta-tech/avb-fix

Port of security or crash fixes from MySQL 5.7.33 to 5.7.34

kamil-holubicki committed:
Merge pull request #4426 from tempesta-tech/avb-fix2

Port of security or crash fixes (MySQL 5.7.34 ... 5.7.35) to 5.6.51

zmur committed:
Bug#33284372 MySQL 5.7 (and Cluster 7.4, 7.5, 7.6) build fails on macos with XCode

Hide bzero macro defined in system headers.

Change-Id: I5132bbee491817d801de96dff6f8fdb0dcc19338

zmur committed:
Bug#33284372 MySQL 5.7 (and Cluster 7.4, 7.5, 7.6) build fails on macos with XCode

XCode includes VERSION file from MySQL source code root instead of C++
language header <version>.

Using MySQL source code root as include path is only used in
storage/perfschema, removing that usage make build succeed.

Change-Id: If7d97dca9ebd14068e7eff1c3171812b4075b9ec

bkandasa committed:
Raise version number after cloning 5.6.54

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.36

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

patrickbirch committed:
PS-7888 Document 5.6 Release Notes (5.6)

        modified:   source/conf.py
        new file:   source/release-notes/Percona-Server-5.6.51-92.0.rst
        modified:   source/release-notes/release-notes_index.rst

wikigurram committed:
Bug#29268656: BOGUS CONSTANT, RANDOM OR TIMEZONE-DEPENDENT
              EXPRESSIONS IN (SUB)PARTITIONING FU

Description
===========
Creating a table with non-deterministic functions in the
generated column expression, should not be allowed as per
the documentation, but this was not being enforced.

Additionally, creating such a table with partitions whose
expression did not actually refer to the problematic column,
failed with a misleading error message about the partition
expression.

Analysis
========
During create table statement parsing, if there are any
unsafe expressions defined for a generated column then
thd->safe_to_cache_query will be set to false. This variable
is later checked while parsing the partition expression.
This check will report an error even if the partition
expression does not refer to the problematic generated
column expression.

Fix
===
Reset the thd->safe_to_cache_query value before parsing the
partition function.

Change-Id: I56e62cd3553db3773b8d0144555ac9a7660224c0

Commit:
Bug #29882299: ALTER USER ... IDENTIFIED WITH ... BY ... SHOULD BE A PRIVILEGED
               OPERATION

Problem: user when tries to change plugin, CREATE USER privilege is not checked.
Fix: Check if IDENTIFIED WITH clause is specified or not when connected user
changes the plugin for his own account. Note: this is backport of bug27923149

RB#27015

phulakun committed:
Bug#33141958 - THE FIRST ASAN UAF ISSUE OF MYSQL SERVER

Statements which commits a transaction implicitly or explicitly are
not allowed inside a trigger and a stored function. CREATE TRIGGER
and CREATE FUNCTION reports an error in this case.

DROP TABLESPACE statement implicitly commits a transaction. But
while creating a trigger and a stored function, this statement is
allowed by mistake. Because of which, assertion to make sure
statement is allowed inside a trigger or a stored function, fails
while executing it in the debug build. An error is reported in the
non-debug build.

To fix this issue, included DROP TABLESPACE statement in the
list of statements which are not allowed inside a trigger and a
stored function. CREATE TRIGGER and CREATE FUNCTION reports an
error if body contains DROP TABLESPACE statement.

Change-Id: I18b89ae4b63e7b1b40d5b3961a7b9eb9711f849c

Commit:
Bug#33238010: partition table access second index lead to cpu high load

The mysql_template of m_prebuilt (and some flags) seems always same among each partitions during same query.
build_template() calls are 1 at first of query is optimal.

RB: 26729
Reviewed-by: Debarun Banerjee <debarun.banerjee@oracle.com>

Commit:
Bug#33417058: The fuction buf_validate() affects performance seriously and can be optimized

NOTE: Performance problem only on debug build.

buf_pool_validate_instance() is called for each BP instances.
But it calls buf_LRU_validate(), all BP instance validation. (for all instances)
It should call buf_LRU_validate_instance().

RB: 27073
Reviewed-by: Marcin Babij <marcin.babij@oracle.com>

Commit:
Bug #33468647 Backport cmake code from Bug#30359287 and Bug#28170711 to 5.7

Comments in cmake/ssl.cmake for 5.7 say:

  For Windows or macOS, WITH_SSL="system" is handled a bit differently:
  We assume you have installed
      https://slproweb.com/products/Win32OpenSSL.html
      We look for "C:/Program Files/OpenSSL-Win64/"
      The .dll files must be in your PATH.

Backport cmake code from
    Bug #30359287 CMAKE THINKS PERL IS SYSTEM SSL ON WINDOWS
    Bug #28170711 ADD CMAKE SUPPORT FOR OPENSSL 1.1 ON WINDOWS

This allows us to build with headers and libraries
from slproweb, ignoring files from  e.g. Strawberry Perl.

Look for static libraries, so that we don't have to extend PATH.

Change-Id: I45ffb0a26074484c02dccf5cccb231266125cd24

pedrolgomes committed:
BUG#33213841: MySQL crashes while executing `RESET REPLICA ALL`

When the replica SQL mode includes the mode 'PAD_CHAR_TO_FULL_LENGTH',
the channel name can be read from the repository with trailing spaces
what can lead to errors in operations that use that name to identify
the channel.

To solve this issue when reading from replication repo table we unset
the mode "PAD_CHAR_TO_FULL_LENGTH"

ReviewBoard: 27047
ReviewBoard: 27008

Commit:
Bug#33141966 - INCONSISTENT BEHAVIOR IF A COLUMN OF TYPE SERIAL IS SET
AS GENERATED

A column of type SERIAL expands to BIGINT UNSIGNED NOT NULL
AUTO_INCREMENT UNIQUE. Defining a column as both SERIAL and GENERATED
ALWAYS causes undefined behaviour.

Fix:
Do not allow a column of type SERIAL to be a generated column.

Change-Id: I4299726e891d7609f5cfc2dc6d1298fa53d9b0b8

patrickbirch committed:
PXC-3775 Checking incompatibilities with Netlify and Jenkins (5.6)

modified:   Makefile
	new file:   netlify.toml
	new file:   requirements.txt
	new file:   runtime.txt
	new file:   source/conf-netlify/conf.py
	modified:   source/ext/psdom.py
	new file:   source/psdom.py

patrickbirch committed:
Merge pull request #1506 from patrickbirch/pxc-5.6-3775

PXC-3775 Checking incompatibilities with Netlify and Jenkins (5.6)

patrickbirch committed:
PXC-3778 Additional settings for Netlify (5.6)

modified:   source/conf-netlify/conf.py

patrickbirch committed:
Merge pull request #1509 from patrickbirch/pxc-5.6-3778

PXC-3778 Additional settings for Netlify (5.6)

karthik-kamath committed:
BUG#28142052: BACKPORT FIX FOR BUG 24293015 TO 5.7

The problem was that SHOW PROCESSLIST could end up reading
freed memory when accessing THD::m_query_string belonging to
a connection that was in the process of deleting a
Prepared_statment (e.g. due to reprepare).

The reason was that THD::m_query_string could be allocated
on the wrong MEM_ROOT - not the statement MEM_ROOT, but the
MEM_ROOT belonging to the prepared statement. So that even
if SHOW PROCESSLIST locks the mutex protecting
THD::m_query_string, the prepared statement could make the
string invalid by deleting
Prepared_statement::main_mem_root.

This patch fixes the problem by making sure that
THD::m_query_string is allocated on the correct memroot
before the prepared statement terminates.

The patch also includes some minor refactoring related to
'expanded_query'.

Note:
This is a backport of the patch for Bug#24293015 from 8.0 to
5.7.

Change-Id: Idd3a074d932afd5e672898490847f83a76236300

salmankhan123417 committed:
BUG#32803050 REPLICA WITH FEDERATED TABLE CRASHES IN HA_FEDERATED::RND_POS WHEN APPLYING ROW EVENTS

Description: When executing Update/Delete with indexes on a table with
Row based replication setup and the federated table on slave, The Slave
crashes.

Analysis: Slave crashes when trying to get the reference to the row
to be update or deleted on federated table using position() method.
The crash occurs due to assert(stored_result) in position method, since
the stored_result is still empty. Server should do index_scan and
update the stored_result and then call for position() and ha_rnd_pos()
methods to get the ref to the row stored. But in this scenario the
server calls for position() method without scanning the table resulting
in crash at assert.

Fix: The fix is added to fetch the position or reference of the row
only when stored_result is not empty.

RB: 26755

gkodinov committed:
Bug #31933415: OUT OF BOUND READ IN ER()
RB#27195

Fixed the case of unknown error.

Commit:
Bug#32989716: MISSING CORRECT CLOSE SESSION HANDLING DURING
              KEYRING MIGRATION

Description: Keyring migration server uses a mysql
connection to perform online migration. This mysql session
is not closed gracefully after the migration. This leads to
a 'Note' being printed on the online server error log:-
"[Note] Aborted connection x to db: 'xyz' user: 'abc' host:
'localhost' (Got an error reading communication packets)"

Analysis: After the online keyring migration, migration
server process is aborted before gracefully closing the
mysql session. This happens because the migration server
process exits before calling the destructor of
'Migrate_keyring', where the closing of the mysql session
is handled.

Fix: This issue is fixed in 8.0 as part of a code
refactoring done in this area by WL#13446. Back-porting the
relevant code changes to 5.7 to fix this bug and to make
the code consistent across versions.

RB#27186

patrickbirch committed:
Merge pull request #4458 from patrickbirch/ps-5.6-7888

PS-7888 Document 5.6 Release Notes (5.6)

patrickbirch committed:
PS-7973 Fix date in 5.6 RN (5.6)

	modified:   source/release-notes/Percona-Server-5.6.51-92.0.rst

patrickbirch committed:
Merge pull request #4530 from patrickbirch/ps-5.6-7973

PS-7973 Fix date in 5.6 RN (5.6)

Commit:
Bug #33362907 mysql crashes due to small thread_stack value

Add a check_stack_overrun() call in PT_option_value_list::contextualize(),
since we may call super::contextualize() too late.

Change-Id: If92ff658ca47c755f08e2b67e3e665d3ec7ce2ae
(cherry picked from commit db62015b75d60b8bb4b6f7235a51423c899de123)

Commit:
Bug#31374305 - FORMAT() NOT DISPLAYING WHOLE NUMBER SIDE CORRECTLY FOR
ES_MX AND ES_ES LOCALES

Changed the grouping and decimal separator for spanish locales as per
ICU.

Change-Id: I5d80fa59d3e66372d904e17c22c532d4dd2c565b

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 1]

Patch 1: git add the original tarball.

https://thrysoee.dk/editline/libedit-20210910-3.1.tar.gz

Remove before checkin: config.guess config.sub, m4/lt~obsolete.m4
m4/lt-obsolete.m4 contains illecal characters '~'

Change-Id: Ifcc1a110699f372cebeb7d2689359cc1bb5ef6e3
(cherry picked from commit 0a84fde76c48c64031c8eb4873327344b94368b6)
(cherry picked from commit 11a943ec744a53c76b2e8189e545ae0580f2c71d)

wikigurram committed:
Bug#32975221: 5.7.29+ REGRESSION, SPATIAL INDEX NOT USED
              ANYMORE FOR MBRCONTAINS

Description:
============
This is a regression from the fix for Bug#29770705. The fix
introduced cloning of range tree for a special case when we
have a predicate having an AND condition ("pred1 AND pred2")
and one part of this AND condition is not a simple
predicate(an OR condition - "pred2"). For such a case, we
needed to clone the range tree created for "pred1". However,
when we clone the range tree, we do not copy the "geometry
flag" associated with the range tree. This results in
ignoring the spatial index available to do the range scan.

Fix
===
Copy the `rkey_func_flag` from original tree if GEOM_FLAG is
set.

On 8.0, this issue is not reproducible because the use_count
handling is different and clone() is not invoked. However it
is clear that for some case it might fail. So we need the
changes, hence added the unit test for the same.

Change-Id: I36312bc2c11db4e4e77d67e9443a98b37ade38e8

avbelov23 committed:
BUG#17447086 - FTS: DUPLICATE FTS_DOC_ID VALUE ON TABLE

It's a race condiction in fts_get_next_doc_id and was not completely
fixed in rb://2171(bug#16469399).

rb://5523 approved by Jimmy.Yang

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 2]

Patch 2: Add cmake code based on the .ac and .am autotools files.

No relevant changes to .ac and .am files, so copy
  src/CMakeLists.txt
  src/config.h

Rename src/makelist to src/makelist.in so we can subst @AWK@ in it.
Change path to current libedit in cmake/readline.cmake.

Change-Id: I4f4c25a2194a2f434f4e1ec5680ca4116d77c4d9
(cherry picked from commit 7c962babc74fab81c07c091606b5be64d2fdc36a)
(cherry picked from commit 86c5543769f198c80d98e35eb7b94abdfda23193)

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 3]

Patch 3:

Add patch for:
Bug #31360025 UPDATE TO LIBEDIT 20190324-3.1 BREAKS CONTROL-C BEHAVIOUR IN MYSQL CLIENT
handle SIGINT in read_char()

No need to apply patch for:
Bug#11757855 49967: BUILT-IN LIBEDIT DOESN'T READ .EDITRC ON LINUX
upstream ifdefs have been cleaned up.

Change-Id: Id1ae08ff213be0bc3c5d957b4cf2840fd00aee07
(cherry picked from commit 660a8d1c29592c70f3a2fef65cad66490bde3b28)
(cherry picked from commit 741abce3ac26b4f049712921657d91ec00f5509b)

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 4]

Patch 4:

Remove old version: extra/libedit/libedit-20191231-3.1

Change-Id: Ied89826b8c74c2e1d47c9e3287577a0fef6d0b0d
(cherry picked from commit 31a598a1acfe7818350daa8f2ff07e8cf2371cbf)
(cherry picked from commit 293d2a8c03cde51a0f7a7914829eeeb0e43366f2)

avbelov23 committed:
BUG#31716706 BAD HANDLING OF QUOTES IN SHOW GRANTS

Description: SHOW GRANTS quotes username with ' and if the username
contains ' it is not escaped. When the output of SHOW GRANTS is
dumped to a file to re-excecute the commands will result in either
invalid SQL or SQL injection.

Analysis: It is hardcoded to quote the username with ' and also fails
to escaping the ' when present in username. Right way is to append the
quotes and escape it with \ when quotes are present in username or host.

Fix: Escape the ' with \ if present in username@host specification.

RB: 26845

avbelov23 committed:
Bug#27082268 Invalid FTS sync synchronization

The fix closes two issues:
Bug #27082268 - INNODB: FAILING ASSERTION: SYM_NODE->TABLE != NULL DURING FTS SYNC
Bug #27095935 - DEADLOCK BETWEEN FTS_DROP_INDEX AND FTS_OPTIMIZE_SYNC_TABLE

Both issues were related to a FTS cache sync being done during
operations that perfomed DDL actions on internal FTS tables
(ALTER TABLE, TRUNCATE). In some cases the FTS tables and/or
internal cache structures could get removed while still being
used to perform FTS synchronization leading to crashes. In other
the sync operations could not get finishes as it was waiting for
dict lock which was taken by thread waiting for the background
sync to be finished.

The changes done includes:
- Stopping background operations during ALTER TABLE and TRUNCATE.
- Removal of unused code in FTS.
- Cleanup of FTS sync related code to make it more readable and
easier to maintain.

RB#18262

avbelov23 committed:
Bug #32831765 SERVER HITS OOM CONDITION WHEN LOADING TWO INNODB

TABLES WITH FTS INDEXES

Problem:
Concurrent INSERTs into the tables (with FTS index) causes huge
number of FTS_SYNC requests are sent and rapid growth in memory
usages.
When concurrent inserts into single table, FTS_OPT thread get SYNC
request when fts cache threshold is reached. In this case FTS_OPT
thread fetches the SYNC request and performs FTS SYNC operation.
When concurrent inserts into many tables, SYNC requests are sent for
many tables. At a time, FTS_OPT thread can process one SYNC request.
In the meantime, each INSERT into other table keeps sending SYNC
request to FTS_OPT thread.

Fix:
1. added a variable total_size_before_sync in fts cache to record total
size of fts cache when SYNC request is sent.
2. next SYNC request is sent based on the delta fts cache size
increment from last SYNC request.
3. reset total_size_before_sync when SYNC completes.

RB: #26571
Reviewed by : Rahul Agarkar <rahul.agarkar@oracle.com>

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [Solaris]

Additional patch, fixing build break on Solaris.

Solaris has MIN in <utility.h> and <sys/sysmacros.h>
(rather than <sys/param.h>)

Change-Id: I1e319e6d580758097e495f58fd1fa7bc8e91a08b
(cherry picked from commit 798b9a207238b3218e506666efd07d168c11fd05)

avbelov23 committed:
Bug#32552332 - CAN'T WRITE; DUPLICATE KEY IN TABLE '/TMP/#SQL...'

In the case of a query that uses temporary table for aggregation,
the group-by Item is used as the unique constraint of the temporary
table.

If the group-by Item value is already present, the row is updated.
Otherwise, a new row is inserted into the temporary table.

If the group-by Item is an Item with a result field(like Item_func)
or a reference Item, it gets evaluated twice. Once to check if the
result exists in the temporary table and if not, again while
constructing the row to insert.

If the group-by item is non-deterministic, the result value used to
check for existence would be different from the value with which an
insert is tried. This causes the insert to fail if the value already
exists in the table.

Fix:

If any of the group-by Items are non-deterministic, use the hash of
the group-by Items as the unique constraint of the temporary table as
the hash gets evaluated only once.

avbelov23 committed:
BUG#32965864 MULTI-THREADED UNSAFE ACCESS TO A FUNCTION LOCAL STATIC …

…VARIABLE

Description
-----------
Both `MYSQL_BIN_LOG::report_missing_purged_gtids` and
`MYSQL_BIN_LOG::report_missing_gtids` have an output parameter name
`errmsg` that will, under given conditions, be instantiated with a
non-thread-local function-local static variable reference.

Invocation of both methods may occur from different threads execution
stacks and concurrent changes to the static variable may occur.

Side effects include threads logging an error message intended for other
threads or a scrambled error message.

Fix
---
Use `std::string` as the output parameter type, instead of `const char**`
and remove the `static` modifier from the function-local variable.

Reviewed-by: Justin Jose <justin.jose@oracle.com>
Reviewed-by: Pedro Gomes <pedro.gomes@oracle.com>
RB: 26731

bkandasa committed:
Updated copyright year in user visible text

    Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Updated copyright year in user visible text

    Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

satya-bodapati committed:
PS-7929 : Wrong row lock number in SEIS

https://jira.percona.com/browse/PS-7929

Problem:
--------

The number of row locks in "SHOW ENGINE INNODB STATUS" OUTPUT is wrong.
When the trx inserts many duplicate records, the number of row locks
in, "SHOW ENGINE INNODB STATUS" output is wrong.

InnoDB maintains this locks as a counter in trx->lock.n_rec_locks. They are
incremented on lock acquisition and decremented on lock release.

In case of duplicate records, some lock acquisitions increment the counter
for the lock which is already acquired/set

Fix

Commit:
PS-7993: Use dot to separate MTR test combination name in JUnit report

https://jira.percona.com/browse/PS-7993

Using underscore as a separator between test name and test combination
name makes it impossible for any script used to parse results file to
find out if there is a test combination in test name. Use dot as
a separator in this case to improve this.
Added junit_combinations test suite to be able to verify this behavior.

Commit:
PS-7995: Update macOS version used for Azure builds to latest

https://jira.percona.com/browse/PS-7995

The macOS-10.14 environment is going to be deprecated soon.
Use macOS-10.15 instead.

oleksandr-kachan committed:
Merge pull request #4562 from oleksandr-kachan/PS-7995-5.7

PS-7995: Update macOS version used for Azure builds to latest (5.7)

venkatesh-prasad-v committed:
Merge pull request #4474 from venkatesh-prasad-v/5.7-PS-7792

PS-7792: On replica Master_User should be blank if not storing in slave_master_info

EvgeniyPatlan committed:
Merge pull request #4573 from percona/release-5.7.36-39

Release 5.7.36 39

patrickbirch committed:
Merge pull request #4382 from patrickbirch/ps-5.7-7779

PS-7779 Document system-wide crypto-policies for RHEL/CentOS (5.7)

patrickbirch committed:
PS-7965 Release notes for 5.7.36-39 (5.7) (#4528)

modified:   source/conf.py
	new file:   source/release-notes/Percona-Server-5.7.36-39.rst
	modified:   source/release-notes/release-notes_index.rst

patrickbirch committed:
PS-8010 Remove ext from Makefile (5.7) (#4571)

Modified: Makefile

AgileStas committed:
[PS-8007] fix: Start mysqld service after local/remote fs targets

oleksandr-kachan committed:
Merge pull request #4556 from oleksandr-kachan/PS-7993-5.7

PS-7993: Use dot to separate MTR test combination name in JUnit report (5.7)

AgileStas committed:
Merge pull request #4579 from AgileStas/PS-8007_fs-require-57

[PS-8007] fix: Start mysqld service after local/remote fs targets

Commit:
PS-8020 Update Percona logo in PDF
modified: /source/conf.py new file: /source/percona-logo-color.png

alina-derkach-oaza committed:
Merge pull request #4583 from alina-derkach-oaza/ps-5.7-8020

PS-8020 Update Percona logo in PDF (5.7)

kamil-holubicki committed:
PS-8028: Server crashes with Assertion `m_prebuilt->table->n_ref_count > 0' failed

https://jira.percona.com/browse/PS-8028

Cause:
5.7 does not support atomic DDL. It may happen that MySql's frm file is
out of sync with InnodDB dictionary. In such a case the table that is
part of InnoDB partitioned table is not opened which causes asserts.

Fix:
We are not able to recover from this case automatically, but we can
improve the error handling.
If the partitioned table cannot be opened, and its 1st partition is
marked as corrupted, return error without proceeding with further
opening.

Also, logging was improved as it may happen that when the inconsistency
is detected (opening the table), partition names are not initialized
yet, which results with the spoiled log.

voodam committed:
Crash when auto incremented column updated in partitioned table

https://jira.percona.com/browse/PS-7856

Use dbug_tmp_use_all_columns/dbug_tmp_restore_column_map as described in 74cc73d as a workaround, because autoincremented column is not used in the query.

kamil-holubicki committed:
Merge pull request #4541 from tempesta-tech/avb-backport

Backport of security or crash fixes (MySQL 5.7.35 ... 5.7.36) to 5.6

satya-bodapati committed:
Merge pull request #4542 from satya-bodapati/PS-5.7-7929

PS-7929 : Wrong row lock number in SEIS

venkatesh-prasad-v committed:
Merge pull request #4592 from venkatesh-prasad-v/5.7-PS-8030

[5.7] PS-8030: Multithreaded replica crashes inside Relay_log_info::cannot_safely_rollback()

kamil-holubicki committed:
Merge pull request #4587 from kamil-holubicki/PS-8028

PS-8028: Server crashes with Assertion `m_prebuilt->table->n_ref_coun…

Commit:
PS-8048 : Implemented (Merge MySQL 5.7.37) (docs)

https://jira.percona.com/browse/PS-8048

***
Updated man pages from MySQL Server 5.7.37 source tarball.
***
Updated 'scripts/fill_help_tables.sql' from MySQL Server 5.7.37 source tarball.

kamil-holubicki committed:
Merge pull request #4615 from percona/release-5.6.51-92.0

Release 5.6.51 92.0

avbelov23 committed:
PS-7890 : Unable to install RocksDB engine plugin if server is started
with --loose-rocksdb_persistent_cache_size_mb

Added checks to bypass RocksDB restrictions

kamil-holubicki committed:
PXC-3878: Merge PS 5.6 backport security fixes

Merge with conflicts.

# Conflicts:
#	sql/sys_vars.cc

kamil-holubicki committed:
PXC-3878: Merge PS 5.6 backport security fixes

Merge conflicts resolved.

kamil-holubicki committed:
Merge pull request #4601 from tempesta-tech/avb-PS-7890-rocksdb-persistent-cache-size-5.7

PS-7890 : Unable to install RocksDB engine plugin if server is started with --loose-rocksdb_persistent_cache_size_mb

kamil-holubicki committed:
Merge pull request #4589 from tempesta-tech/am-PS-7856

Crash when auto incremented column updated in partitioned table

kamil-holubicki committed:
Merge pull request #1584 from kamil-holubicki/PXC-3878

PXC 3878 - Merge PS 5.6 backport security fixes

rahulmalik87 committed:
PS-8032 Inplace Add index with lock=exclusive does not generate MLOG_ADD_INDEX redo
https://jira.percona.com/browse/PS-8032
Problem:
Server is not generating MLOG_ADD_INDEX redo in case of inplace add index
with exclusive lock DDL. MLOG_ADD_INDEX tell if any transaction
happened without redo logging. It is very important for backup tool to know if
any transaction happened without redo logging.

Fix:
Generate MLOG_ADD_INDEX if index was created using btr_blk.

patrickbirch committed:
Merge pull request #1462 from patrickbirch/pxc-5.6-3699

PXC-3699 wsrep_cluster_name documentation wrong about dynamic option …

Commit:
Implemented PS-8048 (Merge MySQL 5.7.37 up to d773112b3a819360faa71023b147c0ecd8bc90e6) (merge with conflicts)

https://jira.percona.com/browse/PS-8048

Merge commit 'd773112b3a819360faa71023b147c0ecd8bc90e6' into merge-mysql-5.7.37-part1

Commit:
PS-8048: Merge MySQL 5.7.37 part 1 (up to d773112b) (conflicts resolved)

https://jira.percona.com/browse/PS-8048
***
VERSION raised to "5.7.37-40.0".
PERCONA_INNODB_VERSION in univ.i raised to "40.0".

Commit:
PS-8048: Merge MySQL 5.7.37 part 1 (up to d773112b) (Fixed Azure failures)

https://jira.percona.com/browse/PS-8048
***
Fixed Microsoft Azure failures

rahulmalik87 committed:
Merge pull request #4595 from rahulmalik87/5.7-PS-8032

5.7 PS-8032 Inplace Add index with lock=exclusive doesn’t generate MLOG_A…

Commit:
PS-8048: Merge MySQL 5.7.37 part 1 (up to d773112b) (Fixed test failures)

https://jira.percona.com/browse/PS-8048
***
Fixed test failures:
Added debug only check in partition_fmt_inconsistency.test
Fixed test case mysql_config_editor.test by recording correct licence
Fixed unittests by making part of keys in
ascending order in CloneSpatialKey test of opt_range-t.cc

Commit:
PS-8048: Merge MySQL 5.7.37 part 1 (up to d773112b) (Changed CLUSTERING_FLAG id to 1 << 29)

https://jira.percona.com/browse/PS-8048
***
Changed CLUSTERING_FLAG id to 1 << 29
Kept SERIAL_FLAG id as previous value of (1<< 28)

nitendra-bhosle committed:
Merge pull request #4614 from nitendra-bhosle/merge-mysql-5.7.37-part1

PS-8048: Merge MySQL 5.7.37 part 1 (up to d773112b)

Commit:
PXC-3908: Merge PS 5.7.37 Part 1 with conflicts

https://jira.percona.com/browse/PXC-3908

Merge commit 'cd5e5fe7269d42c65f29c18c2619e7ad67161655' into 5.7.37-psmerge-part-1

Commit:
PXC-3908: Merge PS 5.7.37 Part 1 (Conflicts resolved)

https://jira.percona.com/browse/PXC-3908

Changed MYSQL_VERSION_PATCH version from 36 to 37,
MYSQL_VERSION_EXTRA version from 39 to 40
in MYSQL_VERSION file

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1605) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1605) before we can accept your contribution.<br/>**3** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: patrickbirch<br/>:white_check_mark: kamil-holubicki<br/>:white_check_mark: nitendra-bhosle<br/>:x: Nitendra Bhosle<br/>:x: rahulmalik87<br/><hr/>**Nitendra Bhosle** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtradb-cluster?pullRequest=1605) it.</sub>

nitendra-bhosle commented:
Jenkins : https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/283/

Commit:
PXC-3908: Merge PS 5.7.37 Part 1 (Fixed failing tests and unittest failures)

https://jira.percona.com/browse/PXC-3908

Fixed below test failures :

galera.galera_as_slave_crash_safe (Fixed by
Adding MASTER_USER= 'root' in testcase file )

innodb.alter_table_stage_progress ( re-recored
test case , as extra information is logged)

Fixed unittest failures by making cmake changes,
mdl test case in unittest depends on wsrep library,
but wsrep library only builts when WITH_WSREP
macro is defined so changed unittest/gunit/CMakeLists.txt
file to reflect this.

nitendra-bhosle commented:
Updated jenkins : https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/287/

venkatesh-prasad-v commented on sql/rpl_mi.cc:
```diff
@@ -404,7 +404,7 @@ bool Master_info::read_info(Rpl_info_handler *from)
                      (ulong) BIN_LOG_HEADER_SIZE) ||
       from->get_info(host, sizeof(host), (char *) 0) ||
-      from->get_info(user, sizeof(user), (char *) 0) ||
+      from->get_info(user, sizeof(user), (char *) "test") ||
```
This change re-introduces the bug that was fixed in commit efd3968b98fd6affe5b582da34f4a7fc95c546e0.
Can you please revert this change and check if something can be fixed in test itself?

venkatesh-prasad-v reviewed (approved):
LGTM

Commit:
PXC-3908: Null merge of PXC-5.6

https://jira.percona.com/browse/PXC-3908

Merge remote-tracking branch 'percona-pxc/5.6' into 5.7.37-psmerge-part-1

kamil-holubicki reviewed (approved):
lgtm
